homeassistant/components/habitica/config_flow.py

```diff
@@ -0,0 +1,84 @@
+"""Config flow for habitica integration."""
+import logging
+
+from aiohttp import ClientResponseError
+from habitipy.aio import HabitipyAsync
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import CONF_API_KEY, CONF_NAME, CONF_URL
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import CONF_API_USER, DEFAULT_URL, DOMAIN  # pylint: disable=unused-import
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_USER): str,
+        vol.Required(CONF_API_KEY): str,
+        vol.Optional(CONF_NAME): str,
+        vol.Optional(CONF_URL, default=DEFAULT_URL): str,
+    }
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect."""
+
+    async def authenticate(username, password, url=DEFAULT_URL) -> bool:
+        websession = async_get_clientsession(hass)
+        api = HabitipyAsync(conf={"login": username, "password": password, "url": url})
+        try:
+            await api.user.get(session=websession)
+        except ClientResponseError:  # pylint: disable=broad-except
+            return False
+        return True
+
+    is_valid = await authenticate(
+        username=data[CONF_API_USER], password=data[CONF_API_KEY], url=data[CONF_URL]
+    )
+    if is_valid is False:
+        raise InvalidAuth()
+    return {
+        "title": f"{data.get('name', 'Default username')}",
+        CONF_API_USER: data[CONF_API_USER],
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for habitica."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+
+        errors = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except InvalidAuth:
+                errors = {"base": "invalid_credentials"}
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors = {"base": "unknown"}
+            else:
+                await self.async_set_unique_id(info[CONF_API_USER])
+                self._abort_if_unique_id_configured()
```

Comment:
Please add a test where we test that two config entries can't be setup with the same username. Add a `MockConfigEntry` with a user in the data and then start the flow and try to add the same user.

https://github.com/home-assistant/core/blob/165dd351b7bdfa55d56d0ee18d97d9b101ccce19/tests/common.py#L657

Reply:
@MartinHjelmare, do you have any good test example you could share/reference so i could knock that last test? 👍 :)

Reply:
Here's an example:
https://github.com/home-assistant/core/blob/c8ac787c9f7be9059dfefdf937bcb270f87aa226/tests/components/hue/test_config_flow.py#L145-L175

Reply:
Thank you, @MartinHjelmare. Finally had some time to finish it up, please take a look